server/master/src/main/java/org/apache/accumulo/master/Master.java

```diff
@@ -1586,6 +1586,11 @@ public void update(LiveTServerSet current, Set<TServerInstance> deleted,
       cleanListByHostAndPort(serversToShutdown, deleted, added);
     }
 
+    // clear out any servers that are no longer current
+    // this is needed when we are using a fate operation to shutdown a tserver as it
+    // will continue to add the server to the serversToShutdown (ACCUMULO-4410)
+    serversToShutdown.retainAll(current.getCurrentServers());
```

Comment:
I was looking at how this `update()` method is called an noticed that its conditionally called when the set of tservers changes.  This made me think that could be still be race conditions where junk lingers in `serversToShutdown`.

This race condition made me wonder if `Master.StatusThread.updateStatus()` could be more defensive.   I thought this would be a really simple change but realized the need to synchronize complicated it.   Below is the result of my poking at the code.  I don't like the complexity below, wish I could just call `Sets.intersect().isEmpty()` directly in the if stmt.

```java
    private long updateStatus()
        throws AccumuloException, AccumuloSecurityException, TableNotFoundException {
      Set<TServerInstance> currentServers = tserverSet.getCurrentServers();
      tserverStatus = gatherTableInformation(currentServers);
      checkForHeldServer(tserverStatus);

      // See ACCUMULO-XYZ
      boolean stsEmpty;
      Collection<TServerInstance> stsView = null;
      synchronized (serversToShutdown) {
        //need to sync to avoid concurent mod exception while iterating (impl of Sets.intersection iterates)
        SetView<TServerInstance> intersection = Sets.intersection(serversToShutdown, currentServers);
        stsEmpty = intersection.isEmpty();
        if(!stsEmpty) {
          stsView = new ArrayList<>(intersection);
        }
      }

      if (!badServers.isEmpty()) {
        log.debug(
            "not balancing because the balance information is out-of-date " + badServers.keySet());
      } else if (notHosted() > 0) {
        log.debug("not balancing because there are unhosted tablets: " + notHosted());
      } else if (getMasterGoalState() == MasterGoalState.CLEAN_STOP) {
        log.debug("not balancing because the master is attempting to stop cleanly");
      } else if (!stsEmpty) {
        log.debug("not balancing while shutting down servers " + stsView);
      } else {
        for (TabletGroupWatcher tgw : watchers) {
          if (!tgw.isSameTserversAsLastScan(currentServers)) {
            log.debug("not balancing just yet, as collection of live tservers is in flux");
            return DEFAULT_WAIT_FOR_WATCHER;
          }
        }
        return balanceTablets();
      }
      return DEFAULT_WAIT_FOR_WATCHER;
    }
```

I also thought of doing `serversToShutdown.retainAll(currentServers)` in `Master.StatusThread.updateStatus()` instead of using `Sets.intersect()`.  However this just feels wrong to me.   I like the idea of `Master.StatusThread.updateStatus()` only reading these sets.

Reply:
I suppose we could do this in addition to the retainAll in the update() call as a safety, but I question whether the additional synchronization is worth it.

Reply:
Updated the LiveTServerSet to call the Master.update(...) method everytime, and pushing the check for changes into that method.  This will ensure that the serversToShutdown is cleaned up and avoids having to add this piece of code.